### PR TITLE
Initial per user implementation

### PR DESCRIPTION
**Pull request by danderson00**
Simple implementation. It loads the target record on each operation. Any thoughts on optimizations?

Again, we need to be careful that this is secure in the default configuration. More test cases are probably required. Thoughts?

As far as our discussion on customizing container names, the user is welcome to do that, I don't think there is any additional code needed here. We need to make sure we make people aware of the potential security hole when the do customise container names. I also added https://github.com/Azure/azure-mobile-apps-node-files/issues/18 - it allows customising container names with a user id.

Resolves https://github.com/Azure/azure-mobile-apps-node-files/issues/16

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
     "jshint": "^2.9.1",
     "mocha": "^2.2.5",
     "sinon": "^1.17.4",
+    "sqlite3": "^3.1.4",
```
Comment (danderson00) [inline package.json]: Required for integration tests

File: src/index.js [hunk 2]
```diff
@@ -4,17 +4,17 @@
 
 var blobStorage = require('./blobStorage'),
     filesRouter = require('./router'),
-    storage = require('azure-storage');
+    azureStorage = require('azure-storage');
```
Comment (danderson00) [inline src/index.js]: This was actually a bug

File: src/index.js [hunk 3]
```diff
@@ -4,17 +4,17 @@
...
 
-    if(!(storageConfiguration.connectionString || (storageConfiguration.account && storageConfiguration.key))))
+    if(!(storageConfiguration.connectionString || (storageConfiguration.account && storageConfiguration.key)))
         throw new Error('No Azure storage account information was provided. Please configure a Data Connection or configure manually. See http://azure.github.io/azure-mobile-apps-node/module-azure-mobile-apps_src_storageConfiguration_Environment.html for more information.');
 
-    var blobClient = storage.createBlobService(storageConfiguration.connectionString || storageConfiguration.account, storageConfiguration.key),
+    var blobClient = azureStorage.createBlobService(storageConfiguration.connectionString || storageConfiguration.account, storageConfiguration.key),
         storage = blobStorage(blobClient, storageConfiguration),
-        router = filesRouter(storageConfiguration, storage, logger);
+        router = filesRouter(configuration, storage, logger);
```
Comment (danderson00) [inline src/index.js]: So was this. The perUser tests actually exercise this index module now. I should probably add a vanilla integration test as well.

File: src/perUser.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    return function (req, res, next) {
+        var context = req.azureMobile,
+            id = req.params.id,
+            userId = context.user && context.user.id,
+            tableDefinition = configuration.tables[req.params.tableName],
+            table = context.tables(req.params.tableName);
+
+        if(tableDefinition && tableDefinition.perUser) {
+            table.where({ id: id }).read().then(function (items) {
+                if(items.length === 1 && userId === items[0][configuration.userIdColumn])
```
Comment (danderson00) [inline src/perUser.js]: Added https://github.com/Azure/azure-mobile-apps-node/issues/460 to add find method to data provider to return a single result. This is a clunky experience for server side code.

Comment (mamaso) [inline src/perUser.js]: Any reason why we aren't filtering with the userIdColumn in the query object?
```js
var query = { id: id }; 
query[configuration.userIdColumn] = userId;
table.where(query).read().then(function (items) {
  if(items.length) ... else ...
```

If only we had full es6, dynamic keys in object literals :/

Comment (danderson00) [inline src/perUser.js]: Yeh, this is neater, I like it. Been thinking about ES6. I wonder if we have anyone even using node <4.

**Comment by danderson00** [comment]
As far as optimiZations go, there is a possible way to avoid the hit on the table each request - if we customize the container name to include the user id, unauthorized users simply won't see the files, but they aren't stopped from adding files to ANY record. Not ideal.

**Comment by danderson00** [comment]
Resolves https://github.com/Azure/azure-mobile-apps-node-files/issues/18

File: test/blobStorage.tests.js [hunk 5]
```diff
@@ -15,28 +15,28 @@ describe('blobStorage', function () {
...
+        return storage.list('table-1').then(function (list) {
+            expect(list.entries.length).to.equal(1);
         });
     });
 
-    it('container name is set from resolver', function () {
-        configuration.containerResolver = function (table, id) {
-            return table + '_' + id;
-        };
+    // it('container name is set from resolver', function () {
```
Comment (mamaso) [inline test/blobStorage.tests.js]: why is this commented out :P

Comment (danderson00) [inline test/blobStorage.tests.js]: Because it's no longer relevant here. I meant to remove it. It's now removed. Cheers!

**Comment by danderson00** [comment]
Resolves https://github.com/Azure/azure-mobile-apps-node-files/issues/18

**Pull request merged**